### PR TITLE
refactor gum_x86_writer_put_jcc_* and gum_exec_block_virtualize_branch_insn to use capstone ids instead of opcode magic handling

### PR DESCRIPTION
**Pull request by erik-smit**
refactor gum_x86_writer_put_jcc_* and gum_exec_block_virtualize_branch_insn to use capstone ids instead of opcode magic handling

to fix processing of unusual prefixes like "REPNE JB"
0xF2 0x72 0x02 REPNE JB

fixes current failing test /Core/Stalker/follow_repne_jb

### CODE REVIEW

File: gum/backend-x86/gumstalker-x86.c [hunk 1]
```diff
@@ -31,7 +31,7 @@
 
 #define GUM_CODE_ALIGNMENT                     8
 #define GUM_DATA_ALIGNMENT                     8
-#define GUM_CODE_SLAB_SIZE_IN_PAGES        65536
+#define GUM_CODE_SLAB_SIZE_IN_PAGES         1024 // <ole> Erik Could you try lowering this: https://github.com/frida/frida-gum/blob/master/gum/backend-x86/gumstalker-x86.c#L34
```
Comment (erik-smit) [inline gum/backend-x86/gumstalker-x86.c]: i'm not sure if this belongs in this PR. ;)

File: gum/backend-x86/gumstalker-x86.c [hunk 2]
```diff
@@ -2746,6 +2745,48 @@ gum_cpu_reg_from_capstone (x86_reg reg)
...
+      return X86_INS_JS;
+    case X86_INS_JO:
+      return X86_INS_JNO;
+    case X86_INS_JP:
+      return X86_INS_JNP;
+     case X86_INS_JS:
+      return X86_INS_JNS;
+    default:
+      g_assert_not_reached ();
+  };
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Redundant semicolon.

File: gum/backend-x86/gumstalker-x86.c [hunk 3]
```diff
@@ -1896,10 +1897,8 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
 
       gum_exec_block_close_prolog (block, gc);
 
-      gum_x86_writer_put_jcc_near_label (cw,
-          gum_x86_reader_jcc_opcode_negate (
-              gum_x86_reader_jcc_insn_to_short_opcode (insn->begin)),
-          is_false, GUM_NO_HINT);
+      gum_x86_writer_put_jcc_near_label (cw, gum_negate_jcc(insn->ci->id), is_false,
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: - Missing space before parenthesis.
- Trailing whitespace.
- Exceeds 80 cols.

File: gum/backend-x86/gumstalker-x86.c [hunk 4]
```diff
@@ -1896,10 +1897,8 @@ gum_exec_block_virtualize_branch_insn (GumExecBlock * block,
 
       gum_exec_block_close_prolog (block, gc);
 
-      gum_x86_writer_put_jcc_near_label (cw,
-          gum_x86_reader_jcc_opcode_negate (
-              gum_x86_reader_jcc_insn_to_short_opcode (insn->begin)),
-          is_false, GUM_NO_HINT);
+      gum_x86_writer_put_jcc_near_label (cw, gum_negate_jcc(insn->ci->id), is_false, 
+		  GUM_NO_HINT);
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Hanging indent should be 4 spaces.

File: gum/backend-x86/gumstalker-x86.c (hunk 1 above)
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Would be good to do this in a separate one so I can squash the commits of this one easily. :)

File: gum/backend-x86/gumspinlock-amd64-msc.c [hunk 5]
```diff
@@ -9,6 +9,8 @@
 #include "gumx86writer.h"
 #include "gummemory.h"
 
+#include <capstone.h>
+
```
Comment (oleavr) [inline gum/backend-x86/gumspinlock-amd64-msc.c]: This include should not be necessary, as gumx86writer.h should provide it due to the instruction type being part of its API.

File: gum/arch-x86/gumx86writer.h [hunk 6]
```diff
@@ -102,20 +104,29 @@ enum _GumPtrTarget
...
   /* signed */
-  GUM_X86_JL  = 0x7c,
-  GUM_X86_JGE = 0x7d,
-  GUM_X86_JLE = 0x7e,
-  GUM_X86_JG  = 0x7f,
+  GUM_X86_JL    = 0x7c,
+  GUM_X86_JGE   = 0x7d,
+  GUM_X86_JLE   = 0x7e,
+  GUM_X86_JG    = 0x7f,
 };
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.h]: These should be deleted and inlined instead, or moved to the implementation if used more than once. But assuming they're only used once, the `switch` already explains which instructions it's generating opcodes for, so these enums don't add any clarity.

File: gum/arch-x86/gumx86writer.c [hunk 7]
```diff
@@ -2361,3 +2343,51 @@ gum_x86_writer_put_prefix_for_registers (GumX86Writer * self,
...
+    case X86_INS_JO:
+      return GUM_X86_JO;
+    case X86_INS_JP:
+      return GUM_X86_JP;
+    case X86_INS_JS:
+      return GUM_X86_JS;
+    case X86_INS_JCXZ:
+      return GUM_X86_JCXZ;
+     case X86_INS_JECXZ:
+       return GUM_X86_JECXZ;
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Indent of these two lines off by one space.

File: gum/arch-x86/gumx86writer.c [hunk 8]
```diff
@@ -850,22 +828,26 @@ gum_x86_writer_put_jcc_near (GumX86Writer * self,
...
 
 void
 gum_x86_writer_put_jcc_near_label (GumX86Writer * self,
-                                   guint8 opcode,
+                                   x86_insn instruction_id,
                                    gconstpointer label_id,
                                    GumBranchHint hint)
 {
-  gum_x86_writer_put_jcc_near (self, opcode, GSIZE_TO_POINTER (self->pc), hint);
+  gum_x86_writer_put_jcc_near (self, instruction_id, GSIZE_TO_POINTER (self->pc), hint);
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Exceeds 80 cols.

File: gum/arch-x86/gumx86writer.c [hunk 9]
```diff
@@ -850,22 +828,26 @@ gum_x86_writer_put_jcc_near (GumX86Writer * self,
...
 gum_x86_writer_put_jcc_short_label (GumX86Writer * self,
-                                    guint8 opcode,
+                                    x86_insn instruction_id,
                                     gconstpointer label_id,
                                     GumBranchHint hint)
 {
-  gum_x86_writer_put_jcc_short (self, opcode, GSIZE_TO_POINTER (self->pc),
+  guint8 opcode;
+
+  opcode = gum_get_jcc_opcode (instruction_id);
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Not used.

File: gum/arch-x86/gumx86writer.c [hunk 10]
```diff
@@ -850,22 +828,26 @@ gum_x86_writer_put_jcc_near (GumX86Writer * self,
...
+
+  gum_x86_writer_put_jcc_short (self, instruction_id, GSIZE_TO_POINTER (self->pc),
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Exceeds 80 cols.

File: gum/arch-x86/gumx86writer.c [hunk 11]
```diff
@@ -832,11 +807,14 @@ gum_x86_writer_put_jcc_short (GumX86Writer * self,
 
 void
 gum_x86_writer_put_jcc_near (GumX86Writer * self,
-                             guint8 opcode,
+                             x86_insn instruction_id,
                              gconstpointer target,
                              GumBranchHint hint)
 {
   gssize distance;
+  guint8 opcode;
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Variables should be declared in the order that they are used, except if one of them represents the return value/result, in which case it goes first.

File: gum/arch-x86/gumx86writer.c [hunk 12]
```diff
@@ -832,11 +807,14 @@ gum_x86_writer_put_jcc_short (GumX86Writer * self,
...
+
+  opcode = gum_get_jcc_opcode (instruction_id);
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: I guess we could drop the variable as doing the call inline would be equally clear?

File: gum/arch-x86/gumx86writer.c [hunk 13]
```diff
@@ -784,42 +786,15 @@ gum_x86_writer_put_jmp_near_ptr (GumX86Writer * self,
...
 gum_x86_writer_put_jcc_short (GumX86Writer * self,
-                              guint8 opcode,
+                              x86_insn instruction_id,
                               gconstpointer target,
                               GumBranchHint hint)
 {
   gssize distance;
+  guint8 opcode;
+
+  opcode = gum_get_jcc_opcode (instruction_id);
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: I guess we could drop the variable as doing the call inline would be equally clear?

File: gum/arch-x86/gumx86writer.c [hunk 14]
```diff
@@ -784,42 +786,15 @@ gum_x86_writer_put_jmp_near_ptr (GumX86Writer * self,
...
-
-void
 gum_x86_writer_put_jcc_short (GumX86Writer * self,
-                              guint8 opcode,
+                              x86_insn instruction_id,
                               gconstpointer target,
                               GumBranchHint hint)
 {
   gssize distance;
+  guint8 opcode;
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Variables should be declared in the order that they are used, except if one of them represents the return value/result, in which case it goes first.

File: gum/arch-x86/gumx86relocator.c [hunk 15]
```diff
@@ -467,7 +467,7 @@ gum_x86_relocator_rewrite_conditional_branch (GumX86Relocator * self,
       gconstpointer is_true = GSIZE_TO_POINTER (unique_id | 1);
       gconstpointer is_false = GSIZE_TO_POINTER (unique_id | 0);
 
-      gum_x86_writer_put_jcc_short_label (ctx->code_writer, 0xe3, is_true,
+      gum_x86_writer_put_jcc_short_label (ctx->code_writer, ctx->insn->id, is_true,
```
Comment (oleavr) [inline gum/arch-x86/gumx86relocator.c]: Exceeds 80 cols.

File: gum/arch-x86/gumx86relocator.c [hunk 16]
```diff
@@ -479,7 +479,7 @@ gum_x86_relocator_rewrite_conditional_branch (GumX86Relocator * self,
     else
     {
       gum_x86_writer_put_jcc_near (ctx->code_writer,
-          gum_x86_reader_jcc_insn_to_short_opcode (ctx->start),
+          ctx->insn->id,
           target,
           GUM_NO_HINT);
```
Comment (oleavr) [inline gum/arch-x86/gumx86relocator.c]: I think we can unwrap this a bit:

```C
      gum_x86_writer_put_jcc_near (ctx->code_writer, ctx->insn->id, target,
          GUM_NO_HINT);
```

File: gum/backend-x86/gumstalker-x86.c [hunk 17]
```diff
@@ -338,6 +338,7 @@ static void gum_write_segment_prefix (uint8_t segment, GumX86Writer * cw);
 
 static GumCpuReg gum_cpu_meta_reg_from_real_reg (GumCpuReg reg);
 static GumCpuReg gum_cpu_reg_from_capstone (x86_reg reg);
+static x86_insn gum_negate_jcc (x86_insn instruction_id);;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Extra semicolon.

File: gum/backend-x86/gumstalker-x86.c [hunk 18]
```diff
@@ -2746,6 +2745,48 @@ gum_cpu_reg_from_capstone (x86_reg reg)
...
+    case X86_INS_JNP:
+      return X86_INS_JP;
+    case X86_INS_JNS:
+      return X86_INS_JS;
+    case X86_INS_JO:
+      return X86_INS_JNO;
+    case X86_INS_JP:
+      return X86_INS_JNP;
+     case X86_INS_JS:
+      return X86_INS_JNS;
```
Comment (oleavr) [inline gum/backend-x86/gumstalker-x86.c]: Indent of these two lines off by one space.

File: gum/arch-x86/gumx86writer.h (hunk 6 above)
Comment (erik-smit) [inline gum/arch-x86/gumx86writer.h]: They used to be used in a few places, which is why I just used them also.
But all their uses get replaced in this PR by capstone-ids, so I can remove and fold it into gum_generate_opcode() now.

Comment (oleavr) [inline gum/arch-x86/gumx86writer.h]: Perfect!

File: gum/arch-x86/gumx86relocator.c [hunk 19]
```diff
@@ -467,8 +467,8 @@ gum_x86_relocator_rewrite_conditional_branch (GumX86Relocator * self,
       gconstpointer is_true = GSIZE_TO_POINTER (unique_id | 1);
       gconstpointer is_false = GSIZE_TO_POINTER (unique_id | 0);
 
-      gum_x86_writer_put_jcc_short_label (ctx->code_writer, 0xe3, is_true,
-          GUM_NO_HINT);
+      gum_x86_writer_put_jcc_short_label (ctx->code_writer, ctx->insn->id,
```
Comment (oleavr) [inline gum/arch-x86/gumx86relocator.c]: Trailing whitespace.

File: gum/arch-x86/gumx86writer.c [hunk 20]
```diff
@@ -850,22 +822,23 @@ gum_x86_writer_put_jcc_near (GumX86Writer * self,
...
 void
 gum_x86_writer_put_jcc_short_label (GumX86Writer * self,
-                                    guint8 opcode,
+                                    x86_insn instruction_id,
                                     gconstpointer label_id,
                                     GumBranchHint hint)
 {
-  gum_x86_writer_put_jcc_short (self, opcode, GSIZE_TO_POINTER (self->pc),
-      hint);
+  gum_x86_writer_put_jcc_short (self, instruction_id,
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: Trailing whitespace.

File: gum/arch-x86/gumx86writer.c [hunk 21]
```diff
@@ -2361,3 +2334,51 @@ gum_x86_writer_put_prefix_for_registers (GumX86Writer * self,
...
+      return 0x7D;
+    case X86_INS_JLE:
+      return 0x7E;
+    case X86_INS_JG:
+      return 0x7F;
+    case X86_INS_JCXZ:
+      return 0xE3;
+    case X86_INS_JECXZ:
+      return 0xE3;
+    case X86_INS_JRCXZ:
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: These three cases can be merged.

File: gum/arch-x86/gumx86writer.c [hunk 22]
```diff
@@ -2373,10 +2373,8 @@ gum_get_jcc_opcode (x86_insn instruction_id)
     case X86_INS_JG:
       return 0x7F;
     case X86_INS_JCXZ:
-      return 0xE3;
-    case X86_INS_JECXZ:
-      return 0xE3;
     case X86_INS_JRCXZ:
+    case X86_INS_JECXZ:
```
Comment (oleavr) [inline gum/arch-x86/gumx86writer.c]: - Would be good to preserve the order like it was.
- Hexdecimal constants should be lower-case for consistency.

**Pull request merged**